### PR TITLE
#491: s3 improvments

### PR DESCRIPTION
**Pull request by efgefg0001**
1. flat_json, csv, json format + gzip compression - COMPLETED
2. s3 configurator ui + backend - COMPLETED

### CODE REVIEW

File: configurator/frontend/src/catalog/destinations/lib/s3.tsx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      type: stringType
+    },
+    {
+      id: '_formData.s3Bucket',
+      displayName: 'Bucket',
+      required: true,
+      type: stringType
+    },
+    {
+      id: '_formData.s3Region',
```
Comment (xtreding) [inline configurator/frontend/src/catalog/destinations/lib/s3.tsx:109]: Please add drop-down select of the `region` field. Get an example from `Redshift` destination configuration.

File: configurator/frontend/src/catalog/destinations/lib/s3.tsx [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+);
+
+const destination = {
+  description: <>
+    S3 Object Storage is ideal for backups and to archive company data. It is a convenient, affordable and compliant way to store any amount of static data
+  </>,
+  syncFromSourcesStatus: 'not_supported',
+  id: 's3',
+  type: 'database',
+  displayName: 'S3',
```
Comment (xtreding) [inline configurator/frontend/src/catalog/destinations/lib/s3.tsx:79]: please use displayName = 'Amazon S3'

File: configurator/frontend/src/catalog/destinations/lib/s3.tsx [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      type: selectionType(['flat_json', 'csv'], 1)
+    },
+    {
+      id: '_formData.s3CompressionEnabled',
+      displayName: 'Enable gzip compression',
+      required: false,
+      type: booleanType,
+      defaultValue: false,
+      documentation: <>
+        Enables gzip compression
```
Comment (xtreding) [inline configurator/frontend/src/catalog/destinations/lib/s3.tsx:141]: Let's rewrite documentation like:
`If enabled - all files with events will be compressed (gzip) before uploading to S3. All files will have the suffix '.gzip'`

**Review by xtreding — CHANGES_REQUESTED** [review]

File: server/adapters/s3.go [hunk 4]
```diff
@@ -65,34 +79,66 @@ func NewS3(s3Config *S3Config) (*S3, error) {
 	}
 	s3Session := session.Must(session.NewSession())
 
-	return &S3{client: s3.New(s3Session, awsConfig), config: s3Config}, nil
+	return &S3{client: s3.New(s3Session, awsConfig), Config: s3Config}, nil
 }
 
 //UploadBytes creates named file on s3 with payload
 func (a *S3) UploadBytes(fileName string, fileBytes []byte) error {
```
[the remark this answers is not in the document]
Comment (efgefg0001) [reply server/adapters/s3.go:86]: Refactoring is not correct. I need original bytes for content type detection. Uploaded bytes have original file content type and gzip content encoding.

File: server/adapters/s3.go [hunk 5]
```diff
@@ -47,7 +59,9 @@ func (s3c *S3Config) Validate() error {
 	if s3c.Region == "" {
 		return errors.New("S3 region is required parameter")
 	}
-
+	if s3c.Format == "" {
+		s3c.Format = FLAT_JSON
```
[the remark this answers is not in the document]
Comment (efgefg0001) [reply server/adapters/s3.go]: Moved format field assignment in constructor function.

File: server/adapters/s3.go [hunk 6]
```diff
@@ -16,24 +17,35 @@ import (
...
+	Folder      string           `mapstructure:"folder" json:"folder,omitempty" yaml:"folder,omitempty"`
+	Format      S3EncodingFormat `mapstructure:"format" json:"format,omitempty" yaml:"format,omitempty"`
+	Compression S3Compression    `mapstructure:"compression" json:"compression,omitempty" yaml:"compression,omitempty"`
 }
 
+type S3EncodingFormat string
+type S3Compression string
+
+const (
+	FLAT_JSON S3EncodingFormat = "flat_json"
```
[the remark this answers is not in the document]
Comment (efgefg0001) [reply server/adapters/s3.go]: Renamed constants

File: server/adapters/s3.go [hunk 7]
```diff
@@ -16,24 +17,35 @@ import (
 
 //S3 is a S3 adapter for uploading/deleting files
 type S3 struct {
-	config *S3Config
+	Config *S3Config
```
[the remark this answers is not in the document]
Comment (efgefg0001) [reply server/adapters/s3.go]: Fixed

File: configurator/frontend/src/catalog/destinations/lib/s3.tsx (hunk 2 above)
Comment (efgefg0001) [reply configurator/frontend/src/catalog/destinations/lib/s3.tsx:79]: changed displayName

File: configurator/frontend/src/catalog/destinations/lib/s3.tsx (hunk 3 above)
Comment (efgefg0001) [reply configurator/frontend/src/catalog/destinations/lib/s3.tsx:141]: changed documentation

File: configurator/frontend/src/catalog/destinations/lib/s3.tsx [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      required: false,
+      defaultValue: '',
+      type: stringType
+    },
+    {
+      id: '_formData.s3Format',
+      displayName: 'Format',
+      required: true,
+      defaultValue: 'flat_json',
+      type: selectionType(['flat_json', 'csv'], 1)
```
[the remark this answers is not in the document]
Comment (efgefg0001) [reply configurator/frontend/src/catalog/destinations/lib/s3.tsx:132]: added `json` without flattening, added documentation

File: configurator/frontend/src/catalog/destinations/lib/s3.tsx (hunk 1 above)
Comment (efgefg0001) [reply configurator/frontend/src/catalog/destinations/lib/s3.tsx:109]: Coppied `s3Credentials` component from `Redshift`

File: configurator/frontend/src/catalog/destinations/lib/s3.tsx [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    {
+      id: '_formData.s3Region',
+      displayName: 'Region',
+      required: true,
+      type: stringType
+    },
+    {
+      id: '_formData.s3Endpoint',
+      displayName: 'Endpoint',
+      required: true,
```
[the remark this answers is not in the document]
Comment (efgefg0001) [reply configurator/frontend/src/catalog/destinations/lib/s3.tsx]: removed `endpoint` input

**Pull request merged**